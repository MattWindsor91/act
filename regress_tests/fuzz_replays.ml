(* The Automagic Compiler Tormentor

   Copyright (c) 2018, 2019, 2020 Matt Windsor and contributors

   ACT itself is licensed under the MIT License. See the LICENSE file in the
   project root for more information.

   ACT is based in part on code from the Herdtools7 project
   (https://github.com/herd/herdtools7) : see the LICENSE.herd file in the
   project root for more information. *)

(** Tests the fuzz trace replayer on various pairs of litmus test and trace.
    The test tracks every individual trace in the directory, and derives the
    corresponding test by stripping all extensions and adding '.litmus'. The
    output is the fuzzed output, and the intent is to track unwanted changes
    in action replays. *)

open Core

let litmus_path_of (trace_path : Fpath.t) : Fpath.t =
  trace_path |> Fpath.rem_ext ~multi:true |> Fpath.add_ext "litmus"

let run_on_file ~(file : Fpath.t) ~(path : Fpath.t) : unit Or_error.t =
  ignore file ;
  let litmus_path = litmus_path_of path in
  Or_error.Let_syntax.(
    let%bind trace = Act_fuzz.Trace.load (Plumbing.Input.of_fpath path) in
    let aux = Act_fuzz_run.Filter.Aux.Replay.make trace in
    Act_fuzz_run.Filter.Replay.run aux
      (Plumbing.Input.of_fpath litmus_path)
      Plumbing.Output.stdout)

let run (test_dir : Fpath.t) : unit Or_error.t =
  let full_dir = Fpath.(test_dir / "fuzz" / "replays" / "") in
  Common.regress_on_files "Fuzz replay" ~dir:full_dir ~ext:"trace"
    ~f:run_on_file

let command =
  Common.make_regress_command
    ~summary:"runs litmus observation-parsing regressions" run

let () = Command.run command
