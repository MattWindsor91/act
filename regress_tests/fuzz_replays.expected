# Fuzz replay tests

## fetch.insert-boundary.trace

```
C Fetch

{ x = 42; }

void
P0(atomic_int *x)
{ if (atomic_fetch_add_explicit(x, 1, memory_order_seq_cst) < 0) {  } }

forall (true)

/*
Labels
  {}
Vars
  x: atomic_int*, =?, @global, generated, [Dep;Write]
*/
```
## fetch.insert-negated-addend.trace

```
C Fetch

{ x = 42; }

void
P0(atomic_int *x)
{ if (atomic_fetch_sub_explicit(x, 64, memory_order_seq_cst) < 64) {  } }

forall (true)

/*
Labels
  {}
Vars
  x: atomic_int*, =?, @global, generated, [Dep;Write]
*/
```
## morisset.1.trace

```
C Morisset

{ g1 = 1; g2 = 0; }

void
P0(int *g1, int *g2)
{ *g2 = 42; }

void
P1(int *g1, int *g2)
{
    int l = 0;
    for (l = 0; l <= 4; l++)
    { if (*g1) { break; } for (*g2 = 0; *g2 >= 26; (*g2)++) {  } }
}

forall (g2 == 42)

/*
Labels
  {}
Vars
  g1: int*, =?, @global, existing, [Dep]
  g2: int*, =?, @global, existing, []
  1:l: int, =?, @P1, generated, [Dep]
*/
```
## sbsc.1.trace

```
C SBSC

{
    Xv8jOk6fGbC = 2147483647;
    cQpyaUbj5gR = -521;
    UcoE3gzvYko = 203657823;
    u = -29491608;
    cP = 13;
    BNFKW = -123;
    Is9Qk0n4Rhat = 1720645492;
    kx0hFmyY7lL6 = -42175;
    cQHO4Qc5Vo4 = -2147483648;
    x = 0;
    y = 0;
}

void
P0(int *BNFKW, atomic_int *Is9Qk0n4Rhat, atomic_int *UcoE3gzvYko,
   atomic_int *Xv8jOk6fGbC, int *cP, atomic_bool *cQHO4Qc5Vo4,
   atomic_bool *cQpyaUbj5gR, bool *kx0hFmyY7lL6, int *u, atomic_int *x,
   atomic_int *y)
{
    int a = 0;
    atomic_store_explicit(x, 1, memory_order_relaxed);
    a = atomic_load_explicit(y, memory_order_relaxed);
}

void
P1(int *BNFKW, atomic_int *Is9Qk0n4Rhat, atomic_int *UcoE3gzvYko,
   atomic_int *Xv8jOk6fGbC, int *cP, atomic_bool *cQHO4Qc5Vo4,
   atomic_bool *cQpyaUbj5gR, bool *kx0hFmyY7lL6, int *u, atomic_int *x,
   atomic_int *y)
{
    int a = 0;
    atomic_store_explicit(y, 1, memory_order_relaxed);
    a = atomic_load_explicit(x, memory_order_relaxed);
    atomic_store_explicit(Is9Qk0n4Rhat, 59774975, memory_order_relaxed);
}

exists (0:a == 0 /\ 1:a == 0)

/*
Labels
  {}
Vars
  BNFKW: int*, =-123, @global, generated, []
  Is9Qk0n4Rhat: atomic_int*, =?, @global, generated, [Write]
  UcoE3gzvYko: atomic_int*, =203657823, @global, generated, []
  Xv8jOk6fGbC: atomic_int*, =2147483647, @global, generated, []
  cP: int*, =13, @global, generated, []
  cQHO4Qc5Vo4: atomic_bool*, =-2147483648, @global, generated, []
  cQpyaUbj5gR: atomic_bool*, =-521, @global, generated, []
  kx0hFmyY7lL6: bool*, =-42175, @global, generated, []
  u: int*, =-29491608, @global, generated, []
  x: atomic_int*, =?, @global, existing, []
  y: atomic_int*, =?, @global, existing, []
  0:a: int, =?, @P0, existing, []
  1:a: int, =?, @P1, existing, []
*/
```

Ran 4 test(s).
